### PR TITLE
Concept identification character level embeddings

### PR DESCRIPTION
**Pull request by teodora99**
This pull request contains:

- added character embeddings for both concept identification and head selection
- added GloVe embeddings in the decoder
- changed the hyper-parameters for the concept identification
- changed the size of the GloVe embeddings ( from 50 to 300)

After Alex's pull request is merged, I'll have to resolve the conflicts.

### CODE REVIEW

File: data_pipeline/glove_embeddings.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 import numpy as np
 import pickle
 
-GLOVE_DATA_FILE = 'glove.6B.50d.txt'
+GLOVE_DATA_FILE = 'glove.6B.300d.txt'
```
Comment (alex-frasie) [inline data_pipeline/glove_embeddings.py]: I'd advice you to revert this as the PR with generalization will cause conflicts for this. It's okay that you have tried it locally as well, but I think it would be easier for you afterwards, as you will only need to change GLOVE_EMB_DIM from config, and that's it.

File: models.py [hunk 2]
```diff
@@ -6,9 +6,45 @@
...
 from data_pipeline.glove_embeddings import get_weight_matrix
 
+# this size came from the highest code of a character that was found in the data set
+# の has the code 12398
+NUMBER_OF_ASCII_CHARACTERS = 12400
+
 #TODO: move this.
 BOS_IDX = 1
 
+class CharacterLevelEmbedding(nn.Module):
```
Comment (alex-frasie) [inline models.py:16]: Would it be a good idea to have this class in a separate file? The models file is pretty crowded as it is..

File: models.py [hunk 3]
```diff
@@ -48,6 +94,11 @@ def forward(self, inputs: torch.Tensor, input_lengths: torch.Tensor):
       embedded_inputs = torch.cat((embedded_inputs, glove_embedded_inputs), dim=-1)
     if self.use_glove and not self.use_trainable_embeddings:
       embedded_inputs = self.glove(inputs)
+    if self.use_character_level_embeddings:
+      # compute character level embeddings
+      character_embedded_inputs = self.character_level_embeddings(character_inputs, character_inputs_lengths)
+      # concatenate embedded_inputs with character lever embeddings on the las dimention
```
Comment (alex-frasie) [inline models.py]: Maybe 'concatenate embedded_inputs with character LEVEL embeddings on the LAST DIMENSION' ?

File: data_pipeline/glove_embeddings.py (hunk 1 above)
Comment (teodora99) [reply data_pipeline/glove_embeddings.py]: I sense there will be some conflicts, changed it back.

File: models.py (hunk 2 above)
Comment (Ranapop) [reply models.py:16]: We will have to refactor this at some point. We can do that once the transformer model is merged as well. Then, we can see how to best split them (not sure one file for just the character level embeddings is the best, maybe one file for embeddings or embeddings and encoding)

File: models.py [hunk 4]
```diff
@@ -6,9 +6,45 @@
 
 from data_pipeline.glove_embeddings import get_weight_matrix
 
+# this size came from the highest code of a character that was found in the data set
+# の has the code 12398
+NUMBER_OF_ASCII_CHARACTERS = 12400
```
Comment (Ranapop) [inline models.py:11]: This is quite large. I don't think we have 12400 characters. Perhaps we can drop less standard characters. But maybe we can leave this for now if it doesn't affect the speed

File: models.py [hunk 5]
```diff
@@ -17,18 +53,28 @@ def __init__(self, input_vocab_size, config: CfgNode, use_bilstm=False, glove_em
...
       bidirectional=use_bilstm)
+    if self.use_character_level_embeddings:
+      self.character_level_embeddings = CharacterLevelEmbedding(config)
 
-  def forward(self, inputs: torch.Tensor, input_lengths: torch.Tensor):
+  def forward(self, inputs: torch.Tensor, input_lengths: torch.Tensor,
+              character_inputs: torch.Tensor=None, character_inputs_lengths: torch.Tensor=None):
     """
     Args:
         inputs (torch.Tensor): Inputs (input seq len, batch size).
```
Comment (Ranapop) [inline models.py:88]: Please add comments for character_inputs and character_input_lengths

File: models.py [hunk 6]
```diff
@@ -108,12 +167,15 @@ def forward(self,
 
 class DecoderClassifier(nn.Module):
 
-  def __init__(self, output_vocab_size:int, config: CfgNode):
+  def __init__(self, output_vocab_size:int, config: CfgNode, use_glove: False):
```
Comment (Ranapop) [inline models.py]: shouldn't use glove be in config? also, if it doesn't have a documentation, please add some to the module

File: models.py [hunk 7]
```diff
@@ -108,12 +167,15 @@ def forward(self,
...
     super(DecoderClassifier, self).__init__()
-    # Classifier input is a concatenation of previous embedding - EMB_DIM,
+    # Classifier input is a concatenation of previous embedding - EMB_DIM + GLOVE_EMB_DIM if use_glove,
     # decoder state - HIDDEN_SIZE and context vector - HIDDEN_SIZE.
+    emb_dim = config.EMB_DIM + 2 * config.HIDDEN_SIZE + config.GLOVE_EMB_DIM if use_glove \
+      else config.EMB_DIM + 2 * config.HIDDEN_SIZE
```
Comment (Ranapop) [inline models.py]: this if would be easier to read if it was not in line. Either expand it or put the two sizes in variables and use a short if

File: models.py [hunk 8]
```diff
@@ -128,21 +190,37 @@ class DecoderStep(nn.Module):
...
-    self.classifier = DecoderClassifier(output_vocab_size, config)
-    self.drop_out = config.DROPOUT_RATE
-    self.use_pointer_generator = config.USE_POINTER_GENERATOR
+    self.classifier = DecoderClassifier(output_vocab_size, config, use_glove=self.use_glove)
+    self.drop_out = config.DECODER_DROPOUT_RATE
     if self.use_pointer_generator:
-        self.p_gen_linear = nn.Linear(config.HIDDEN_SIZE * 3 + config.EMB_DIM, 1, bias=True)
+        emb_dim_pointer_generator = config.EMB_DIM + config.HIDDEN_SIZE * 3 + config.GLOVE_EMB_DIM if self.use_glove \
+          else config.EMB_DIM + config.HIDDEN_SIZE * 3
+        self.p_gen_linear = nn.Linear(emb_dim_pointer_generator, 1, bias=True)
```
Comment (Ranapop) [inline models.py]: this if is hard to read. Can you rewrite it to not use "\"?

File: models.py [hunk 9]
```diff
@@ -128,21 +190,37 @@ class DecoderStep(nn.Module):
...
+      weight_matrix = get_weight_matrix(glove_embeddings, config.GLOVE_EMB_DIM)
+      self.glove.load_state_dict({'weight': weight_matrix})
+      self.glove.weight.requires_grad = False
+
+    emb_dim = config.EMB_DIM + config.HIDDEN_SIZE + config.GLOVE_EMB_DIM if self.use_glove \
+      else config.EMB_DIM + config.HIDDEN_SIZE
+    # Lstm input has size EMB_DIM + HIDDEN_SIZE (will take as input the
+    # previous embedding - EMB_DIM and the context vector - HIDDEN_SIZE).
+    self.lstm_cell = nn.LSTMCell(
+      emb_dim, config.HIDDEN_SIZE)
```
Comment (Ranapop) [inline models.py]: You could have kept glove and character level embeddings in two separate PRs

**Review by Ranapop — CHANGES_REQUESTED** [review]
Overall it looks ok, but it would be good if it didn't have changes besides the character level embeddings changes. I left some comments.

File: models.py [hunk 10]
```diff
@@ -6,9 +6,45 @@
...
+    self.embedding = nn.Embedding(NUMBER_OF_ASCII_CHARACTERS, config.CHAR_EMB_DIM)
+    self.gru = nn.GRU(config.CHAR_EMB_DIM, config.CHAR_HIDDEN_SIZE)
+    self.char_hidden_size = config.CHAR_HIDDEN_SIZE
+
+  def forward(self, inputs: torch.Tensor, input_lengths: torch.Tensor):
+    """
+    Compute the learned embeddings at character level
+    :param inputs: (torch.Tensor): Inputs (word len, input seq len, batch size).
+           input_lengths: (torch.Tensor): Inputs Length (input seq len, batch size)
+    :return:
```
Comment (Ranapop) [inline models.py]: this is in a different format from the other documentations --> please change

File: models.py [hunk 11]
```diff
@@ -17,23 +53,36 @@ def __init__(self, input_vocab_size, config: CfgNode, use_bilstm=False, glove_em
...
     if self.use_glove:
       self.glove = nn.Embedding(len(glove_embeddings.keys()), config.GLOVE_EMB_DIM)
       weight_matrix = get_weight_matrix(glove_embeddings, config.GLOVE_EMB_DIM)
       self.glove.load_state_dict({'weight': weight_matrix})
       self.glove.weight.requires_grad = False
-    emb_dim = config.EMB_DIM + config.GLOVE_EMB_DIM if self.use_glove else config.EMB_DIM
+    emb_dim = config.GLOVE_EMB_DIM + config.CHAR_HIDDEN_SIZE if self.use_glove & self.use_character_level_embeddings \
+      else config.GLOVE_EMB_DIM if self.use_glove\
+      else config.CHAR_HIDDEN_SIZE if self.use_character_level_embeddings \
+      else 0
```
Comment (Ranapop) [inline models.py]: this is very hard to follow, please only use short ifs if it's something short and concise. You may gain a few lines of code like this, but you lose readability

File: models.py [hunk 12]
```diff
@@ -6,9 +6,45 @@
...
+    """
+    Compute the learned embeddings at character level
+    :param inputs: (torch.Tensor): Inputs (word len, input seq len, batch size).
+           input_lengths: (torch.Tensor): Inputs Length (input seq len, batch size)
+    :return:
+    """
+    # word_len x input_seq_len x batch_size x char_emb_dim
+    embedded_inputs = self.embedding(inputs)
+    word_len, input_seq_len, batch_size, char_emb_size = embedded_inputs.shape
+    packable_embedded_inputs = embedded_inputs.view(word_len, input_seq_len * batch_size, char_emb_size)
```
Comment (Ranapop) [inline models.py:39]: maybe add a comment and explain what this view does

File: models.py [hunk 13]
```diff
@@ -6,9 +6,45 @@
...
+    packable_input_lengths = input_lengths.reshape(input_seq_len * batch_size)
```
Comment (Ranapop) [inline models.py:40]: why not use view? do you need reshape?

**Review by Ranapop — CHANGES_REQUESTED** [review]
There are a few small changes that still need to be done. Also, please use shortened ifs more sparingly.

File: models.py (hunk 13 above)
Comment (teodora99) [reply models.py:40]: If am using view I get an exception: RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
After changing it to reshape, it worked.

Comment (Ranapop) [reply models.py:40]: yes, but do you understand what that means? Are you just hiding a bug? In my mind the shape should be good, we should not need reshape. Have you checked that the number of elements is good? Reshape is more expensive and if all works well I don't see why we should need it

File: models.py [hunk 14]
```diff
@@ -6,9 +6,49 @@
...
+           inputs: (torch.Tensor): Inputs (word len, input seq len, batch size).
+           input_lengths: (torch.Tensor): Inputs Length (input seq len, batch size)
+       Returns:
+           output: the embedded input sequence
+    """
+    # word_len x input_seq_len x batch_size x char_emb_dim
+    embedded_inputs = self.embedding(inputs)
+    word_len, input_seq_len, batch_size, char_emb_size = embedded_inputs.shape
+    # view will return a tensor with the same data but with a different shape
+    # the reshape and view are used in order to feed the embedded_inputs and inputs_length to the pack_padded_sequence
```
Comment (Ranapop) [inline models.py:38]: the comments are not explaining much in the context of your code...
I would expect smth like "Use view to go from X shape to Y shape" -- that would be more helpful than just putting something one can find in the documentation but not really explain how you use it and why

**Review by Ranapop — CHANGES_REQUESTED** [review]
You still need to improve the comments (for view&reshape) and investigate a bit more why and if you really need reshape
